### ISSUE TITLE
Helper for tests that require asynchronous setup in the controlled page

### ISSUE THREAD

**Issue by jeffposnick** [GoogleChrome/sw-testing-helpers#38]
In the course of writing https://github.com/GoogleChrome/sw-helpers/pull/1 I've found a few situations where I need to run a test that executes code on the controlled page via `WebDriverexecuteScriptAsync()`, but that script needs to wait until something has happened on the controlled page before it's executed.

To give a concrete example, at https://github.com/GoogleChrome/sw-helpers/pull/1/files#diff-a18887f4e658273661a3d4f9f0f4f364R155, I'm testing whether a master entry for a page with an App Cache manifest is present in the Cache Storage API, but I need to wait until after App Cache manifest helper code on the page has run before performing that test.

I've been working around this by delaying the `WebDriverexecuteScriptAsync()` call via `setTimeout()`, but that's obviously ugly and error prone. I could imagine a runtime library provided by `sw-testing-helpers` creating a global `readyToTest` promise that the code being tested would resolve when setup was complete. That might also require the code being tested to detect when it's being run via WebDriver if we wanted to avoid that promise leaking out into production code.

**Comment by gauntface** [comment]
The way I've been handling this is to use

driver.wait(function() {
    return driver.executeScript(function() {
        return (true or false depending on state);
    });
});

This will loop until you return true. In your case you might want to loop checking for some flag set by your code to say it's finished or to keep looking for a particular value in the cache. This is how the running of mocha tests in a page is managed.

It does feel like a common pattern, but I'm concerned that it'll add a dependency on the test (i.e. code will be added to projects to check if the testing framework is there or not), which is a bad signal to this approach.

**Comment by jeffposnick** [comment]
Yeah, I'm concerned about test code leaking into production code, too. There's a discussion at http://stackoverflow.com/a/26689086/385997 about detecting when your code is being run by Selenium.

I guess what I'm proposing is a way of standardizing that "true or false depending on state" check to avoid each project having to implement their own specific "I'm ready" flag. If it were promise-based that could make it even easier.
